File: src/main/java/me/melijn/jda/utils/TableBuilder.java
package me.melijn.jda.utils;

import java.util.*;

public class TableBuilder {


    private List<String> headerRow = new ArrayList<>();
    private Map<Integer, List<String>> valueRows = new HashMap<>();
    private Map<Integer, Integer> columnWidth = new HashMap<>();
    private List<String> footerRow = new ArrayList<>();
    private boolean split;

    public TableBuilder(boolean shouldSplit) {
        this.split = shouldSplit;
    }

    public TableBuilder setColumns(Collection<String> headerNames) {
        headerRow.addAll(headerNames);
        findWidest(headerNames, 0);
        return this;
    }

    public TableBuilder addRow(Collection<String> rowValues) {
        valueRows.put(valueRows.size(), new ArrayList<>(rowValues));
        findWidest(rowValues, 0);
        return this;
    }

    public TableBuilder setFooterRow(Collection<String> footerNames) {
        footerRow.addAll(footerNames);
        findWidest(footerNames, 0);
        return this;
    }

    private void findWidest(Collection<String> footerNames, int startColumn) {
        int temp = startColumn;
        for (String s : footerNames) {
            if ((columnWidth.getOrDefault(temp, 0)) < s.length()) {
                columnWidth.put(temp, s.length());
            }
            temp++;
        }
    }

    public List<String> build() {
        if (valueRows.values().stream().anyMatch(array -> array.size() > headerRow.size())) {
            throw new IllegalArgumentException("A value row cannot have more values then the header (you can make empty header slots)");
        }

        if (footerRow.size() > headerRow.size()) {
            throw new IllegalArgumentException("A footer row cannot have more values then the header (you can make empty header slots)");
        }


        int maxRowWidth = (columnWidth.size() * 3) - 2;
        for (int i : columnWidth.values()) {
            maxRowWidth += i;
        }

        String lijn = "══════════════════════════════════════════════════════════════════════════════════";
        StringBuilder sb = new StringBuilder();
        List<String> toReturn = new ArrayList<>();
        //firstline
        sb.append("```prolog\n╔═").append(lijn, 0, columnWidth.get(0));
        for (int i = 1; i < headerRow.size(); i++) {
            sb.append("═╦═").append(lijn, 0, columnWidth.get(i));
        }
        sb.append("═╗\n");

        //second
        int nums = 0;
        sb.append("║");
        for (String value : headerRow) {
            sb.append(" ").append(value).append(" ".repeat(50), 0, columnWidth.get(nums++) - value.length()).append(" ║");
        }
        sb.append("\n");

        //third
        sb.append("╠═").append(lijn, 0, columnWidth.get(0));
        for (int i = 1; i < headerRow.size(); i++) {
            sb.append("═╬═").append(lijn, 0, columnWidth.get(i));
        }
        sb.append("═╣\n");

        //main
        for (int i = 0; i < valueRows.size(); i++) {
            int numm = 0;
            if (split && sb.length() + maxRowWidth > 1997 - (footerRow.size() > 0 ? maxRowWidth * 3 : maxRowWidth)) {
                toReturn.add(sb.toString() + "```");
                sb = new StringBuilder();
                sb.append("```prolog\n");
            }
            sb.append("║");
            for (String value : valueRows.get(i)) {
                sb.append(" ").append(value).append(" ".repeat(50), 0, columnWidth.get(numm++) - value.length()).append(" ║");
            }
            sb.append("\n");
        }

        //possible end
        if (footerRow.size() > 0) {
            sb.append("╠═").append(lijn, 0, columnWidth.get(0));
            for (int i = 1; i < footerRow.size(); i++) {
                sb.append("═╬═").append(lijn, 0, columnWidth.get(i));
            }
            sb.append("═╣\n");

            int nume = 0;
            sb.append("║");
            for (String value : footerRow) {
                sb.append(" ").append(value).append(" ".repeat(50), 0, columnWidth.get(nume++) - value.length()).append(" ║");
            }
            sb.append("\n");
        }

        sb.append("╚═").append(lijn, 0, columnWidth.get(0));
        for (int i = 1; i < headerRow.size(); i++) {
            sb.append("═╩═").append(lijn, 0, columnWidth.get(i));
        }
        sb.append("═╝");

        toReturn.add(sb.toString() + "```");

        //less gc
        headerRow = null;
        valueRows = null;
        columnWidth = null;
        footerRow = null;
        return toReturn;
    }

}

File: src/main/java/me/melijn/melijnbot/objects/utils/TableBuilder.kt
package me.melijn.melijnbot.objects.utils

import java.util.*
import kotlin.collections.Collection
import kotlin.collections.HashMap
import kotlin.collections.List
import kotlin.collections.MutableList
import kotlin.collections.MutableMap
import kotlin.collections.set
import kotlin.collections.withIndex


class TableBuilder(private val split: Boolean) {

    private var headerRow: MutableList<String> = ArrayList()
    private var valueRows: MutableMap<Int, List<String>> = HashMap()
    private var columnWidth: MutableMap<Int, Int> = HashMap()
    private var footerRow: MutableList<String> = ArrayList()

    companion object {
        const val LINE = "══════════════════════════════════════════════════════════════════════════════════"
    }

    fun setColumns(headerNames: Collection<String>): TableBuilder {
        headerRow.addAll(headerNames)
        findWidest(headerNames)
        return this
    }

    fun addRow(rowValues: Collection<String>): TableBuilder {
        valueRows[valueRows.size] = ArrayList(rowValues)
        findWidest(rowValues)
        return this
    }

    fun setFooterRow(footerNames: Collection<String>): TableBuilder {
        footerRow.addAll(footerNames)
        findWidest(footerNames)
        return this
    }

    private fun findWidest(footerNames: Collection<String>) {
        for ((temp, s) in footerNames.withIndex()) {
            if (columnWidth.getOrDefault(temp, 0) < s.length) {
                columnWidth[temp] = s.length
            }
        }
    }

    fun build(): List<String> {
        if (valueRows.values.stream().anyMatch { array -> array.size > headerRow.size }) {
            throw IllegalArgumentException("A value row cannot have more values then the header (you can make empty header slots)")
        }

        if (footerRow.size > headerRow.size) {
            throw IllegalArgumentException("A footer row cannot have more values then the header (you can make empty header slots)")
        }


        var maxRowWidth = columnWidth.size * 3 - 2
        for (i in columnWidth.values) {
            maxRowWidth += i
        }


        var sb = StringBuilder()
        val toReturn = ArrayList<String>()
        addTop(sb)
        addRow(sb, headerRow)
        addSplicer(sb)

        //main
        for (i in 0 until valueRows.size) {
            if (split && sb.length + maxRowWidth > 1997 - (if (footerRow.size > 0) maxRowWidth * 3 else maxRowWidth)) {
                toReturn.add("$sb```")
                sb = StringBuilder()
                sb.append("```prolog\n")
            }

            valueRows[i]?.let {
                addRow(sb, it)
            }
        }

        if (footerRow.size > 0) {
            addSplicer(sb)
            addRow(sb, footerRow)
        }
        addBottom(sb)

        toReturn.add("$sb```")

        //less gc
        headerRow.clear()
        valueRows.clear()
        columnWidth.clear()
        footerRow.clear()
        return toReturn
    }

    //╔════╦════╗
    private fun addTop(sb: StringBuilder) {
        sb.append("```prolog\n╔═").append(LINE, 0, columnWidth[0])
        for (i in 1 until headerRow.size) {
            sb.append("═╦═").append(LINE, 0, columnWidth[i])
        }
        sb.append("═╗\n")
    }

    //║    ║    ║
    private fun addRow(sb: StringBuilder, list: List<String>) {
        var nume = 0
        sb.append("║")
        for (value in list) {
            sb.append(" ").append(value).append(" ".repeat(50), 0, columnWidth[nume++]?.minus(value.length)).append(" ║")
        }
        sb.append("\n")
    }


    //╠════╬════╣
    private fun addSplicer(sb: StringBuilder) {
        sb.append("╠═").append(LINE, 0, columnWidth[0])
        for (i in 1 until footerRow.size) {
            sb.append("═╬═").append(LINE, 0, columnWidth[i])
        }
        sb.append("═╣\n")
    }

    //╚════╩════╝
    private fun addBottom(sb: StringBuilder) {
        sb.append("╚═").append(LINE, 0, columnWidth[0])
        for (i in 1 until headerRow.size) {
            sb.append("═╩═").append(LINE, 0, columnWidth[i])
        }
        sb.append("═╝")
    }
}